source/postprocess/crystal_preferred_orientation.cc

```diff
@@ -0,0 +1,816 @@
+          out << *file_contents;
+        }
+
+      out.close ();
+
+      if (tmp_filename != filename)
+        {
+          std::string command = std::string("mv ") + tmp_filename + " " + filename;
+          int error = system(command.c_str());
+
+          AssertThrow(error == 0,
+                      ExcMessage("Could not move " + tmp_filename + " to "
+                                 + filename + ". On processor "
+                                 + Utilities::int_to_string(Utilities::MPI::this_mpi_process (MPI_COMM_WORLD)) + "."));
+        }
+
+      // destroy the pointer to the data we needed to write
+      delete file_contents;
+    }
+
+    template <int dim>
+    std::pair<std::string,std::string>
+    CrystalPreferredOrientation<dim>::execute (TableHandler &statistics)
+    {
+
+      const Particle::Property::Manager<dim> &manager = this->get_particle_world().get_property_manager();
+
+      // Get a reference to the CPO particle property.
+      const Particle::Property::CrystalPreferredOrientation<dim> &cpo_particle_property =
+        manager.template get_matching_property<Particle::Property::CrystalPreferredOrientation<dim>>();
+
+      const unsigned int n_grains = cpo_particle_property.get_number_of_grains();
+      const unsigned int n_minerals = cpo_particle_property.get_number_of_minerals();
+
+      // if this is the first time we get here, set the last output time
+      // to the current time - output_interval. this makes sure we
+      // always produce data during the first time step
+      if (std::isnan(last_output_time))
+        last_output_time = this->get_time() - output_interval;
+
+      // If it's not time to generate an output file or we do not write output
+      // return early.
+      if (this->get_time() < last_output_time + output_interval && this->get_time() != end_time)
+        return std::make_pair("","");
+
+      if (output_file_number == numbers::invalid_unsigned_int)
+        output_file_number = 0;
+      else
+        ++output_file_number;
+
+      // Now prepare everything for writing the output and choose output format
+      std::string particle_file_prefix_master = this->get_output_directory() +  "particle_CPO/particles-" + Utilities::int_to_string (output_file_number, 5);
+      std::string particle_file_prefix_content_raw = this->get_output_directory() +  "particle_CPO/CPO-" + Utilities::int_to_string (output_file_number, 5);
+      std::string particle_file_prefix_content_draw_volume_weighting = this->get_output_directory() +  "particle_CPO/weighted_CPO-" + Utilities::int_to_string (output_file_number, 5);
+
+      const auto &particle_handler = this->get_particle_world().get_particle_handler();
+
+      std::stringstream string_stream_master;
+      std::stringstream string_stream_content_raw;
+      std::stringstream string_stream_content_draw_volume_weighting;
+
+      string_stream_master << "id x y" << (dim == 3 ? " z" : " ") << " olivine_deformation_type" << std::endl;
```

Comment:
currently you create a double space in 2D.

```suggestion
      string_stream_master << "id x y" << (dim == 3 ? " z" : "") << " olivine_deformation_type" << std::endl;
```

Reply:
I don't think this is correct. The z also has a space in front of it.

Reply:
never mind, you are right